### PR TITLE
[test-visibility] Add early flake detection section

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do? What is the motivation?

Let users know about Early Flake Detection, a new feature of Test Visibility that allows you to detect flakiness before it reaches your default branch.

### Merge instructions

- [x] Please merge after reviewing

### Additional notes

See [preview](https://docs-staging.datadoghq.com/juan-fernandez/add-early-flake-detection/tests/early_flake_detection/).

### CODE REVIEW

**Comment by nikita-tkachenko-datadog** [comment]
Could we add a row to the [supported features table](https://docs.datadoghq.com/tests/#supported-features) too?

**Review by nikita-tkachenko-datadog — APPROVED** [review]

File: content/en/tests/_index.md [hunk 1]
```diff
@@ -153,22 +154,22 @@ In order to filter using these configurations tags, [you must create facets for
...
 {{< partial name="whats-next/whats-next.html" >}}
 
 [1]: https://app.datadoghq.com/ci/test-services
 [2]: /continuous_integration/explorer/facets/
-[8]: https://app.datadoghq.com/dashboard/lists
-[9]: https://app.datadoghq.com/notebook/list
-[10]: https://app.datadoghq.com/ci/test-runs
-[11]: /monitors/types/ci/
-[12]: /tests/code_coverage/
+[3]: /intelligent_test_runner/
```
Comment (juan-fernandez) [inline content/en/tests/_index.md:171]: unrelated fix

**Comment by juan-fernandez** [comment]
> Could we add a row to the [supported features table](https://docs.datadoghq.com/tests/#supported-features) too?

done in https://github.com/DataDog/documentation/pull/22556/commits/125da2f03afcb554a2ec53b232a638081836a965

File: content/en/tests/early_flake_detection.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+* Datadog's backend automatically stores unique tests for a given test service. These are **known tests**.
+* Before a test session runs, the Datadog library fetches the list of known tests.
+* If a test is in this list, it's executed normally.
+* If a test is **not** in this list, the Datadog library automatically retries it up to 10 times.
+
+Running a test multiple times helps detecting potential issues. If any of the test attempts fails, Datadog automatically tags it as flaky. You may then choose to block the merge of the feature branch with a [Quality Gate][2].
+
+### Why does running a test multiple times help detect issues?
+
+A [flaky test][3] is a test that exhibits both a passing and failing status across multiple test runs for the same commit. This behavior is normally caused by some random component, such as a race condition caused by an unexpected drop in network speed. Running the test multiple times increases the likelihood of this random component to show up.
```
Comment (anmarchenko) [inline content/en/tests/early_flake_detection.md:39]: I don't understand well enough what "exhibits both a passing and a failing status" means, maybe we could rephrase it a bit easier? Like, "test that fails and passes for the same commit" or "test that randomly fails after some number of retries".

**Review by anmarchenko — APPROVED** [review]

File: content/en/tests/early_flake_detection.md (hunk 2 above)
Comment (juan-fernandez) [reply content/en/tests/early_flake_detection.md:39]: this comes directly from our docs: https://docs.datadoghq.com/glossary/#flaky-test but I agree it's not great. I'll think of something better

Comment (juan-fernandez) [reply content/en/tests/early_flake_detection.md]: another pass in https://github.com/DataDog/documentation/pull/22556/commits/4577adb7496add2f8e7f01b72f91924808c02e8d

**Comment by rtrieu** [comment]
Created DOCS-7771 for editorial review.

File: content/en/tests/early_flake_detection.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+## Overview
+
+Early Flake Detection is Datadog's mechanism to tackle test flakiness. By identifying tests that are newly added and running them multiple times, Datadog can detect flakiness before it reaches the default branch. A study shows that up to [75% of flaky tests][1] can be identified this way.
+
+## How it works
+
+* Datadog's backend automatically stores unique tests for a given test service. These are **known tests**.
+* Before a test session runs, the Datadog library fetches the list of known tests.
+* If a test is in this list, it's executed normally.
+* If a test is **not** in this list, the Datadog library automatically retries it up to 10 times.
```
Comment (KickingTheTV) [inline content/en/tests/early_flake_detection.md:33]: I would just re-iterate that not being on the list makes it "new".

If a test is **not** in this list, the test is considered **new**, and the Datadog library will automatically retry the test up to 10 times.

File: content/en/tests/early_flake_detection.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+Running a test multiple times helps detecting potential issues. If any of the test attempts fails, Datadog automatically tags it as flaky. You may then choose to block the merge of the feature branch with a [Quality Gate][2].
+
+### Why does running a test multiple times help detect issues?
+
+A [flaky test][3] is a test that both passes and fails for the same commit (same code). This behavior is normally caused by some random component, such as a race condition caused by an unexpected drop in network speed. Running the test multiple times increases the likelihood of this random component to show up.
+
+
+### How are known tests defined?
+
+The list of known tests is generated from tests run in default and default-like branches. These are [**Excluded Branches**][4].
```
Comment (KickingTheTV) [inline content/en/tests/early_flake_detection.md:44]: These are Excluded Branches from Early Flake Detection.

Since we have excluded branches in different features, think it's worth distinguishing again what we're talking about

File: content/en/tests/early_flake_detection.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+Early Flake Detection normally works as follows:
+
+* A developer working in a feature branch writes a new test, commits and pushes the changes.
+* This test is retried automatically for this commit and every new commit in the feature branch, until the branch is merged.
+* After the feature branch is merged, that new test is considered part of the known tests for the test service. That test is not considered new anymore.
```
Comment (KickingTheTV) [inline content/en/tests/early_flake_detection.md:50]: "That test is not considered new anymore."
change to, "The test is no longer treated as new."

**Review by KickingTheTV — CHANGES_REQUESTED** [review]
LGTM after making a few little copy edits

File: content/en/tests/early_flake_detection.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+* A developer working in a feature branch writes a new test, commits and pushes the changes.
+* This test is retried automatically for this commit and every new commit in the feature branch, until the branch is merged.
+* After the feature branch is merged, that new test is considered part of the known tests for the test service. The test is no longer treated as new.
+
+{{< img src="continuous_integration/early_flake_detection_commits.png" alt="How Early Flake Detection works in your commits.">}}
+
+### Excluded Branches from Early Flake Detection
+
+The Excluded Branches from Early Flake Detection do not execute Early Flake Detection, that is, Early Flake Detection will never retry tests.
```
Comment (KickingTheTV) [inline content/en/tests/early_flake_detection.md:56]: I think this can just be simplified to "The Excluded Branches will not have any of their tests retried by Early Flake Detection."

**Review by KickingTheTV — CHANGES_REQUESTED** [review]
one little change

File: content/en/tests/early_flake_detection.md (hunk 6 above)
Comment (juan-fernandez) [reply content/en/tests/early_flake_detection.md]: done in https://github.com/DataDog/documentation/pull/22556/commits/483aac9cacc374835986b96b50e09ad2bce0869d

**Review by KickingTheTV — APPROVED** [review]
LGTM

File: content/en/tests/early_flake_detection.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
+---
+title: Early Flake Detection
+kind: documentation
+description: Detect flakiness before it reaches the default branch
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:4]:
```suggestion
description: Detect flakiness before it impacts your default branch.
```

File: content/en/tests/early_flake_detection.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+<div class="alert alert-warning">Early Flake Detection is not available in the selected site ({{< region-param key="dd_site_name" >}}) at this time.</div>
+{{< /site-region >}}
+
+{{< callout url="#" btn_hidden="true" >}}
+Early Flake Detection is in public beta.
+{{< /callout >}}
+
+## Overview
+
+Early Flake Detection is Datadog's mechanism to tackle test flakiness. By identifying tests that are newly added and running them multiple times, Datadog can detect flakiness before it reaches the default branch. A study shows that up to [75% of flaky tests][1] can be identified this way.
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:26]:
```suggestion
Early Flake Detection is Datadog's test flakiness solution. By identifying tests that are newly added and running them multiple times, Datadog can detect flakiness before it reaches the default branch. A study shows that up to [75% of flaky tests][1] can be identified this way.
```

File: content/en/tests/early_flake_detection.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+## How it works
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:28]:
```suggestion
```

File: content/en/tests/early_flake_detection.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+* Datadog's backend automatically stores unique tests for a given test service. These are **known tests**.
+* Before a test session runs, the Datadog library fetches the list of known tests.
+* If a test is in this list, it's executed normally.
+* If a test is **not** in this list, the test is considered **new**, and the Datadog library will automatically retry the test up to 10 times.
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:33]:
```suggestion
* If a test is **not** in this list, the test is considered **new**, and the Datadog library automatically retries the test up to 10 times.
```

File: content/en/tests/early_flake_detection.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+Running a test multiple times helps detecting potential issues. If any of the test attempts fail, Datadog automatically tags it as flaky. You may then choose to block the merge of the feature branch with a [Quality Gate][2].
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:35]:
```suggestion
Running a test multiple times helps detect potential issues. If any of the test attempts fail, Datadog automatically tags it as flaky. You can choose to block the merge of the feature branch with a [Quality Gate][2].
```

File: content/en/tests/early_flake_detection.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+* This test has already run in an excluded branch, such as `staging`, `main` or `preprod`.
+* This test is slower than 5 minutes. There is a mechanism not to run Early Flake Detection on tests that are too slow, since retrying these tests could cause significant delays in CI pipelines.
+
+
+### A test was retried that is not new
+
+If the Datadog library can't fetch the full list of known tests, the Datadog library may retry tests that are not new.
+
+There is a mechanism to prevent this error from slowing down the CI pipeline, but if it happens, contact the support team at [Datadog Help][7].
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:105]:
```suggestion
There is a mechanism to prevent this error from slowing down the CI pipeline, but if it happens, contact [Datadog Support][7].
```

File: content/en/tests/early_flake_detection.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+## Troubleshooting
+
+If you suspect there are issues with Early Flake Detection, go to the [Test Service Settings][6], look for your test service, and click **Configure**. Disable Early Flake Detection by clicking on the toggle.
+
+### A new test is not being retried
+
+This could be caused by a couple of reasons:
+
+* This test has already run in an excluded branch, such as `staging`, `main` or `preprod`.
+* This test is slower than 5 minutes. There is a mechanism not to run Early Flake Detection on tests that are too slow, since retrying these tests could cause significant delays in CI pipelines.
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:98]:
```suggestion
* This test has already run in an excluded branch, such as `staging`, `main`, or `preprod`.
* This test is slower than five minutes. There is a mechanism not to run Early Flake Detection on tests that are too slow, since retrying these tests could cause significant delays in CI pipelines.
```

File: content/en/tests/early_flake_detection.md [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+#### Test
+
+New tests have the `@test.is_new` tag set to `true`.
+
+Additionally, retries for this test have the `@test.is_retry` tag set to `true`.
+
+
+## Troubleshooting
+
+If you suspect there are issues with Early Flake Detection, go to the [Test Service Settings][6], look for your test service, and click **Configure**. Disable Early Flake Detection by clicking on the toggle.
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:91]:
```suggestion
If you suspect there are issues with Early Flake Detection, go to the [Test Service Settings page][6], look for your test service, and click **Configure**. Disable Early Flake Detection by clicking on the toggle.
```

File: content/en/tests/early_flake_detection.md [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+## Configuration
+After you have set up your Datadog library for Test Visibility, configure Early Flake Detection from the [Test Service Settings page][6]:
+
+{{< img src="continuous_integration/early_flake_detection_test_settings.png" alt="Early flake Detection in Test Service Settings.">}}
+
+Click on **Configure** in the Early Flake Detection column to see this modal:
+
+{{< img src="continuous_integration/early_flake_detection_configuration_modal.png" alt="Early flake Detection configuration modal.">}}
+
+In this modal you can toggle the status of Early Flake Detection and add a list of [**Excluded Branches from Early Flake Detection**][4] for that test service.
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:72]:
```suggestion
In this modal, you can toggle the status of Early Flake Detection and add a list of [**Excluded Branches from Early Flake Detection**](#excluded-branches-from-early-flake-detection) for that test service.
```

File: content/en/tests/early_flake_detection.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+There is a mechanism to prevent this error from slowing down the CI pipeline, but if it happens, contact the support team at [Datadog Help][7].
+
+## Further Reading
+
+{{< partial name="whats-next/whats-next.html" >}}
+
+[1]: https://2020.splashcon.org/details/splash-2020-oopsla/78/A-Large-Scale-Longitudinal-Study-of-Flaky-Tests
+[2]: /quality_gates/
+[3]: /glossary/#flaky-test
+[4]: /tests/early_flake_detection/#excluded-branches-from-early-flake-detection
```
Comment (alai97) [inline content/en/tests/early_flake_detection.md:114]:
```suggestion
```

**Review by alai97 — CHANGES_REQUESTED** [review]
Some copy suggestions + an anchor link update!

**Comment by juan-fernandez** [comment]
thanks for the feedback @alai97. Any other feedback?

**Review by alai97 — COMMENTED** [review]
LGTM, pending Brand Design diagrams.

**Review by alai97 — APPROVED** [review]
LGTM
